src/index.js

```diff
@@ -181,58 +182,61 @@ Object.assign(Lock.prototype, {
-        // in any case we do not make a next polling try sooner than after 100ms delay
-        // We might make the call sooner if the key is released manually and we get a notification
-        // from Redis PubSub about it
-        ttlTimer = setTimeout(tryAcquire, Math.max(lock.ttl, 100));
-      });
-    };
-
-    if (waitTtl > 0) {
-      expireLockTimer = setTimeout(() => {
-        expired = true;
-        this._subscribers.removeListener(`${this._namespace}:${id}`, tryAcquire);
-        clearTimeout(ttlTimer);
-        // Try one last time and return whatever the acquireLock returns
-        if (!acquiring) {
-          return tryAcquire();
-        }
-      }, waitTtl);
-    }
-
-    // try to acquire a lock
-    tryAcquire();
+  waitAcquireLock(id, lockTtl, waitTtl) {
+    let expired = false; // flag to indicate that the TTL wait time was expired
+    let acquiring = false; // flag to indicate that a Redis query is in process
+
+    let ttlTimer;
+    let expireLockTimer;
+
+    return new Promise((resolve, reject) => {
+      // A looping function that tries to acquire a lock. The loop goes on until
+      // the lock is acquired or the wait ttl kicks in
+      const tryAcquire = () => {
+        this._subscribers.removeListener(`${this._namespace}:${id}`, tryAcquire); // clears pubsub listener
+        clearTimeout(ttlTimer); // clears the timer that waits until existing lock is expired
+        acquiring = true;
+        asCallback(this.acquireLock(id, lockTtl), (err, lock) => {
+          acquiring = false;
+          if (err) {
+            // stop waiting if we hit into an error
+            clearTimeout(expireLockTimer);
+            return reject(err);
+          }
+          if (lock.success || expired) {
+            // we got a lock or the wait TTL was expired, return what we have
+            clearTimeout(expireLockTimer);
+            return resolve(lock);
+          }
+
+          // Wait for either a Redis publish event or for the lock expiration timer to expire
+          this._subscribers.addListener(`${this._namespace}:${id}`, tryAcquire);
+          // Remaining TTL for the lock might be very low, even 0 (lock expires by next ms)
+          // in any case we do not make a next polling try sooner than after 100ms delay
+          // We might make the call sooner if the key is released manually and we get a notification
+          // from Redis PubSub about it
+          ttlTimer = setTimeout(tryAcquire, Math.max(lock.ttl, 100));
+        });
+      };
+
+      if (waitTtl > 0) {
+        expireLockTimer = setTimeout(() => {
+          expired = true;
+          this._subscribers.removeListener(`${this._namespace}:${id}`, tryAcquire);
+          clearTimeout(ttlTimer);
+          // Try one last time and return whatever the acquireLock returns
+          if (!acquiring) {
+            return tryAcquire();
+          }
+        }, waitTtl);
+      }
+
+      // try to acquire a lock
+      tryAcquire();
+    });
```

Comment:
it feels like there should be a better async/await-native representation for this but that's also effort and this should do the trick

Reply:
yeah i was wondering about that… I guess the repeated calls to `tryAcquire` could be unrolled into a loop maybe